contrib_bots/lib/giveaway.py

```diff
@@ -0,0 +1,215 @@
+            if not self.all_warnings_fired:
+                new_content = ('{}/{} of giveaway time elapsed, {} minutes remaining'
+                               .format(self.numerator, self.denominator, str((old_div(self.giveaway_timeMins, 4))*3)))
+                self.time_update(new_content)
+                self.giveaway_stage = self.giveaway_stage + 1
+
+    def check_time(self):
+        if time.time() > (self.giveaway_start_time + self.giveaway_time):
+            print("Giveaway ended")
+            try:
+                self.giveaway_winner = random.choice(self.giveaway_array)
+                self.winner_index = self.giveaway_array.index(self.giveaway_winner)
+                new_content = 'The giveaway has now ended... The winner is... {}!'.format(self.giveaway_winner)
+                direct_content = """Congratulations!, you recently partook in the giveaway called {} and won!
+                Please contact {} on details as to how to claim your prize.""".format(self.giveaway_name, self.giveaway_creator,)
+                self.giveaway_client.send_message(dict(
+                    type='private',
+                    to=self.giveaway_winner_contact_address[self.winner_index],
+                    content=direct_content,
+                ))
+            except:
+                new_content = 'Giveaway ended: No one entered the giveaway :('
+            if self.giveaway_on:
+                self.giveaway_on = False
+                self.giveaway_client.send_message(dict(
+                    type='stream',
+                    to=self.giveawayStream,
+                    subject=self.giveaway_subject,
+                    content=new_content,
+                ))
+        self.time_warning()
+        threading.Timer(6.0, self.check_time).start() # More efficient to check every 6 seconds instead of 1.
+
+    def usage(self):
+        return '''
+            This plugin will allow for users to create
+            giveaways which people can enter. After a set
+            period of time has passed, a random person will
+            be chosen from the list of random people whom
+            entered the giveaway.
+
+            Syntax:
+            @giveaway command giveaway_name giveaway_time
+
+            Command:
+                - init: initialize a giveaway
+                - enter: enter a giveaway (only works if a
+                giveaway is already running)
+                - exit: exit
+                - cancel
+
+            This plugin was created by Daniel O'Brien
+            as part of Google Code-In 2016-2017
+            '''
+
+    def triage_message(self, message, client):
+        # return True if we want to (possibly) response to this message
+
+        original_content = message['content']
+        if message['display_recipient'] == 'giveawayer':
+            return False
+        return original_content.startswith('@giveaway')
+
+    def handle_message(self, message, client, state_handler):
```

Comment:
It would be great to refactor the code in this function to make the size of this function smaller.